File: 08. Trees/01. Binary Tree/13. Top View/main.cpp
#include <bits/stdc++.h>
using namespace std;

class TreeNode {
public:
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

// BFS Approach for Top View
vector<int> topViewBFS(TreeNode* root) {
    if (!root) return {};

    map<int, int> topNodes;  // {horizontal distance : node value}
    queue<pair<TreeNode*, int>> q; // {node, horizontal distance}
    q.push({root, 0});

    while (!q.empty()) {
        auto [node, hd] = q.front();
        q.pop();

        // If no node is placed at this HD, place the first encountered node
        if (topNodes.find(hd) == topNodes.end()) {
            topNodes[hd] = node->val;
        }

        if (node->left) q.push({node->left, hd - 1});
        if (node->right) q.push({node->right, hd + 1});
    }

    vector<int> result;
    for (auto& x : topNodes) {
        result.push_back(x.second);
    }
    return result;
}

// DFS Approach for Top View
void dfs(TreeNode* node, int hd, int depth, map<int, pair<int, int>>& topNodes) {
    if (!node) return;

    // If this HD is not visited or we found a shallower node, update it
    if (topNodes.find(hd) == topNodes.end() || depth < topNodes[hd].second) {
        topNodes[hd] = {node->val, depth};
    }

    // Recur for left and right subtrees
    dfs(node->left, hd - 1, depth + 1, topNodes);
    dfs(node->right, hd + 1, depth + 1, topNodes);
}

vector<int> topViewDFS(TreeNode* root) {
    if (!root) return {};

    map<int, pair<int, int>> topNodes; // {horizontal distance : {value, depth}}
    dfs(root, 0, 0, topNodes);

    vector<int> result;
    for (auto& x : topNodes) {
        result.push_back(x.second.first);
    }
    return result;
}

// Main function
int main() {
    TreeNode* root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    root->left->right = new TreeNode(4);
    root->left->right->right = new TreeNode(5);
    root->left->right->right->right = new TreeNode(6);

    // BFS Solution
    vector<int> resultBFS = topViewBFS(root);
    cout << "Top View (BFS): ";
    for (int val : resultBFS) cout << val << " ";
    cout << endl;

    // DFS Solution
    vector<int> resultDFS = topViewDFS(root);
    cout << "Top View (DFS): ";
    for (int val : resultDFS) cout << val << " ";
    cout << endl;

    return 0;
}

File: 08. Trees/01. Binary Tree/13. Top View/main.py
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None

# BFS Approach for Top View
def topViewBFS(root):
    if not root:
        return []

    top_nodes = {}
    queue = [(root, 0)]

    while queue:
        node, hd = queue.pop(0)

        if hd not in top_nodes:
            top_nodes[hd] = node.val

        if node.left:
            queue.append((node.left, hd - 1))
        if node.right:
            queue.append((node.right, hd + 1))

    return [top_nodes[hd] for hd in sorted(top_nodes.keys())]

# DFS Approach for Top View
def dfs(node, hd, depth, top_nodes):
    if not node:
        return

    if hd not in top_nodes or depth < top_nodes[hd][1]:
        top_nodes[hd] = (node.val, depth)

    dfs(node.left, hd - 1, depth + 1, top_nodes)
    dfs(node.right, hd + 1, depth + 1, top_nodes)

def topViewDFS(root):
    if not root:
        return []

    top_nodes = {}
    dfs(root, 0, 0, top_nodes)

    return [top_nodes[hd][0] for hd in sorted(top_nodes.keys())]

# Main function
def main():
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.left.right = TreeNode(4)
    root.left.right.right = TreeNode(5)
    root.left.right.right.right = TreeNode(6)

    print("Top View (BFS):", topViewBFS(root))
    print("Top View (DFS):", topViewDFS(root))

    
main()
